[ARM][TableGen][MC] Change the ARM mnemonic operands to be optional for ASM parsing

AlfieRichardsArm opened the pull request:
This changs the way the assembly matcher works for Aarch32 parsing. Previously there was a pile of hacks which dictated whether the CC, CCOut, and VCC operands should be present which de-facto chose if the wide/narrow (or thumb1/thumb2/arm) instruction version were chosen.

This meant much of the TableGen machinery present for the assembly matching was effectively being bypassed and worked around.

This patch makes the CC and CCOut operands optional which allows the ASM matcher operate as it was designed and means we can avoid doing some of the hacks done previously. This also adds the option for the target to allow the prioritizing the smaller instruction encodings as is required for Aarch32.

AlfieRichardsArm commented:
If there is any interest from @llvm/issue-subscribers-backend-mips I noticed the sorting by smallest encoding has effects for the MIPS parsing (thus the inclusion of the option rather than having it on as default to not break tests). I don't know what the correct behavior is for MIPS.

AlfieRichardsArm commented:
Also note https://github.com/llvm/llvm-project/pull/83413 is a prerequisite for this change.

AlfieRichardsArm commented:
This change is intended to have minor/no user facing changes. It allows for the removal of some hacks and makes the parsing of ARM more idiomatic with the way TableGen works.

Additionally, it is part of a larger piece of work to address the numerous flaws with how we handle ".w" and ".n".
For more context see: 
https://github.com/llvm/llvm-project/issues/14175
https://gist.github.com/rprichard/2a601c3dd1b281f953b4e08b5a9361bb

AlfieRichardsArm commented:
MCA test failures seem genuine, it makes assumptions about optional operands that this invalidates. Im investigating now

s-barannikov commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -2019,7 +2058,7 @@ emitConvertFuncs(CodeGenTarget &Target, StringRef ClassName,
       << "                              std::begin(TiedAsmOperandTable)) &&\n";
   CvtOS << "             \"Tied operand not found\");\n";
-  CvtOS << "      unsigned TiedResOpnd = TiedAsmOperandTable[OpIdx][0];\n";
+  CvtOS << "      unsigned TiedResOpnd = TiedAsmOperandTable[*(p + 1)][0];\n";
```
Please check if #81381 resolves this issue.

AlfieRichardsArm replied:
It does not, it breaks the tied operands resolution for me, I run into the same issue I had before of the wrong value being used. I think `*(p + 1)` is correct as offsets should not be applied here?

s-barannikov replied:
> I think *(p + 1) is correct as offsets should not be applied here?

Agreed. It looks like it is not an operand index at all.

s-barannikov replied:
I guess the assert above should also be fixed.
And we shouldn't be calculating OpIdx before the switch because it can lead to OOB access...

AlfieRichardsArm replied:
Fixed the assert. Moved this bug fix to a separate patch (https://github.com/llvm/llvm-project/pull/83588).

The OpIdx calculation is needed for the generated switch cases below so I didnt remove it for now. Im not certain how I would remove it just for this case except for adding a if statement? I cant imagine a circumstance where it generates an OOB access though?

s-barannikov replied:
> I cant imagine a circumstance where it generates an OOB access though?

OOB access can happen in HasOptionalOperands case if `*(p + 1)` (index into TiedAsmOperandTable) is larger than the size of DefaultsOffset array.

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -7290,64 +7304,68 @@ bool ARMAsmParser::ParseInstruction(ParseInstructionInfo &Info, StringRef Name,
     // on the Mnemonic based checking to correctly figure out when to put
     // a k_CondCode operand in the list. If we're trying to match the label
     // version, remove the k_CondCode operand here.
```
De-indent the comment, too.

AlfieRichardsArm replied:
Done

s-barannikov commented on llvm/utils/TableGen/PseudoLoweringEmitter.cpp:
```diff
@@ -88,23 +88,26 @@ unsigned PseudoLoweringEmitter::addDagOperandMapping(
+
       // FIXME: Are the message operand types backward?
-      if (DI->getDef() != Insn.Operands[BaseIdx + i].Rec) {
+      if (DI->getDef() != Insn.Operands[i].Rec) {
```
This does not look right. We do need BaseIdx because this function is called recursively for complex operands, and BaseIdx is the offset of the first flattened sub-operand of a complex operand.
(Side note: it is probably not a good idea to call this function recursively in the first place, it is very difficult to follow the logic.)

AlfieRichardsArm commented on llvm/utils/TableGen/PseudoLoweringEmitter.cpp:
```diff
@@ -196,7 +199,7 @@ void PseudoLoweringEmitter::evaluateExpansion(Record *Rec) {
 
   LLVM_DEBUG(dbgs() << "  Operand mapping:\n");
-  for (unsigned i = 0, e = Insn.Operands.size(); i != e; ++i) {
+  for (unsigned i = 0, e = Dag->getNumArgs(); i != e; ++i) {
```
Ah whoops. This change was submitted in this PR: https://github.com/llvm/llvm-project/pull/81605
I will remove from this one as its unrelated

kosarev commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -1474,7 +1513,7 @@ void AsmMatcherInfo::buildOperandMatchInfo() {
                        << (i - NumOptionalOps);
       }
-      if (Op.Class->IsOptional)
+      if (Op.Class->IsOptional && Op.Class->OptionalShouldOffsetCustomParsers)
```
I struggle to understand why this is necessary. By the look of it, the flag makes the operand effectively non-optional. Can you give an example demonstrating the problem this flag addresses?

AlfieRichardsArm replied:
Yes I'll take a  shot at explaining.

This is addressing the `ParserMethod` functionality for operands, (eg. FPImmOperand was the one causing me issues)
The way this is done is at operand parsing time (`ARMAsmParser::parseOperand` in this case), `MatchOperandParserImpl` uses the Mnemonic and the number of operands in the OperandVector at that point to check if the operand at that index has a custom ParserMethod.

This causes issues when an operand with a custom ParserMethod comes after an optional operand as the index can change and the optional operand resolution isn't done until match time (after this point). This flag is here to say these optional operands will _always_ be populated populate at parse time so that the indexes of the operands with custom parsers can be known. (See ARMAsmParser.cpp:7062 for where we add the optional operands even if not present in tokens and ARMAsmParser.cpp:7158 after parsing where it is optionally removed).

This is a pretty poor solution getting around the fundamental problem that if we want to have operand classes with custom parsers, where the class information is tied to the instruction definition, then parsing needs to be done at match time where the operand tokens and classes are matched up. But this would require a much deeper change to the parsing and matching machinery.

AlfieRichardsArm replied:
The reason for this flag is because I wanted to be explicit that this is what I was doing and this is behavior I am relying on. The inverse is also possible (always omitting optional operands) and should be as simple as flipping this flag.

AlfieRichardsArm replied:
Heres a little more context for the example: 
Snippet of tableget for VMOVv4f32:
```
def VMOVv4f32 { // InstructionEncoding Instruction InstTemplate Encoding InstARM NeonI NDataI N1ModImm
  field bits<32> Inst = { 1, 1, 1, 1, 0, 0, 1, SIMM{7}, 1, Vd{4}, 0, 0, 0, SIMM{6}, SIMM{5}, SIMM{4}, Vd{3}, Vd{2}, Vd{1}, Vd{0}, 1, 1, 1, 1, 0, 1, 0, 1, SIMM{3}, SIMM{2}, SIMM{1}, SIMM{0} };
  field bits<32> Unpredictable = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
  field bits<32> SoftFail = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
  int Size = 4;
  string DecoderNamespace = "NEONData";
  list<Predicate> Predicates = [HasNEON];
  string DecoderMethod = "DecodeVMOVModImmInstruction";
  bit hasCompleteDecoder = 1;
  string Namespace = "ARM";
  dag OutOperandList = (outs QPR:$Vd);
  dag InOperandList = (ins nImmVMOVF32:$SIMM, pred:$p);
  string AsmString = "vmov${p}.f32  $Vd, $SIMM";
  ...
  ```
With an input like:
```
vmoveq.f32 q2, #0.1
```

In this case before this change the parser for `nImmVMOVF32` (which has parser class `FPImmOperand`) was being applied to `Vd` because `${p}` was present in this case and the parser was assuming that it would not be present (the default behaviour previously). 

This flag and always adding _something_ to the OperandVector was the solution to this. 

Also see `Target.td:957` where I try describe this a bit also.

kosarev replied:
Is `vmoveq.f32 q2, #0.1` supposed to be parsed fine with this patch in place? Here I am at `da3d1607ae9a`, and it doesn't seem to work:

```
$ ./llvm-mc -mcpu=cortex-a8 -triple armv7-apple-darwin -debug <<< 'vmoveq.f32 q2, #0.1'
...
<stdin>:1:1: error: invalid instruction, any one of the following would fix this:
vmoveq.f32 q2, #0.1
^
<stdin>:1:16: note: operand must be a register in range [q0, q15]
vmoveq.f32 q2, #0.1
               ^
<stdin>:1:16: note: invalid operand for instruction
vmoveq.f32 q2, #0.1
               ^
```

The immediate still seems to fail to be parsed for `VMOVv4f32`, if that's the intended instruction:
```
Trying to match opcode VMOVv4f32
  Matching formal operand class MCK_CondCode against actual operand at index 1 (<ARMCC::eq>): match success using generic matcher
  Matching formal operand class MCK__DOT_f32 against actual operand at index 2 ('.f32'): match success using generic matcher
  Matching formal operand class MCK_QPR against actual operand at index 3 (<register q2>): match success using generic matcher
  Matching formal operand class MCK_FPImm against actual operand at index 4 (1036831949): operand match failed, recording near-miss with diag code 0
  Matching formal operand class InvalidMatchClass against actual operand at index 5: actual operand index out of range
```

> The way this is done is at operand parsing time (eg. `ARMAsmParser::parseOperand` in this case), where `MatchOperandParserImpl` uses the Mnemonic and the current OperandVector's size to determine if the operand at that index should use a custom ParserMethod.
> 
> This causes issues when an operand with a custom ParserMethod comes after an optional operand as then the index can change.

So how is this a problem, exactly? The `maskTrailingOnes()` call above is supposed to mark all positions at which an operand with a custom parser may appear, taking into account all optional operands that may precede it. Doesn't it do it correctly in your case for some reason?

> In this case before this change the parser for `nImmVMOVF32` (which has parser class `FPImmOperand`) was being applied to `q2` instead of `#0.1` because `${p}` was present in this case and the parser was assuming that it would not be present (the default behaviour previously).

Again, how you think is this wrong? In presence of optional operands, a custom parser being applied to an umatching operand is normal. Such a custom parser is supposed to return no-match, thus letting know the rest of the parsing infrastructure that this is an operand of some other kind and so other parsers should be tried.

AlfieRichardsArm replied:
Sorry, bad example `vmoveq.f32 q0, #0.125` works with `bin/llvm-mc -debug -triple armv8a-none-eabi -mattr=-fullfp16,+neon -show-encoding <<< "vmov.f32 q0, #0.125"`

AlfieRichardsArm replied:
Hmm thank you for your insight here. It definitely wasn't working without this, but perhaps that was a bug in a different area and not what I thought it was. 

I think maybe this bug stems from `ARMAsmParser::parseFPImm` in several situations produces an error rather than `NoMatch`. I will see if I can change this to get around this issue a different way.

AlfieRichardsArm replied:
It turns out this was a symptom of the remaining hacks surrounding VPTPred not being optional ):
I am going to try make this optional also, thank you for catching this

AlfieRichardsArm replied:
@kosarev I have resolved this. I had misdiagnosed an issue where certain custom parsers where behaving poorly when being applied to non-matching operands. The new patch also makes VPT operand optional and fixes these cases, removing the need for this option.

kosarev commented on llvm/include/llvm/Target/Target.td:
```diff
@@ -1724,6 +1733,11 @@ class Target {
+  // SortBySize = Should the assembly matcher prefer the smaller
+  // instructions. 1 if the instruction set should sort by size,
+  // 0 otherwise.
+  int SortBySize = 0;
```
Please address this with a separate patch.

The `Sort` bit in the name sounds like a reference to an implementation detail. Can we just directly say the implied and name it something like `PreferSmallerInstructions`?

The type should be `bit`, not `int`?

AlfieRichardsArm replied:
Just to clarify you would like the SortBySize target option (renamed as requested) and machinery to be in a separate patch? 
I am happy to do this, however the ARM target doesn't work with this turned on without the rest of the patch, so the patch to add this option would not turn it on for the ARM target?

AlfieRichardsArm replied:
This has been pulled out to https://github.com/llvm/llvm-project/pull/83587

kosarev commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -608,12 +627,19 @@ struct MatchableInfo {
-  /// operator< - Compare two matchables.
-  bool operator<(const MatchableInfo &RHS) const {
+  /// comp - Compare two matchables.
+  bool comp(const MatchableInfo &RHS, const CodeGenTarget &Target) const {
```
We could probably use the chance and give it some meaningful name. Something like `shouldBeMatchedBefore()` maybe?

kosarev commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -608,12 +627,19 @@ struct MatchableInfo {
 
+    // Sort by the resultant instuctions size, eg. for ARM instructions
+    // we must choose the smallest matching instruction.
+    if (Target.getSortBySize()) {
+      if (ResInstSize != RHS.ResInstSize)
```
Can be a single `if`. +Same below.

AlfieRichardsArm replied:
Done

AlfieRichardsArm committed:
[ARM][TableGen][MC] Change the ARM mnemonic operands to be optional for ASM parsing

This changs the way the assembly matcher works for Aarch32 parsing.
Previously there  was a pile of hacks which dictated whether the CC
and CCOut operands should be present which de-facto chose if the
wide/narrow (or thumb1/thumb2/arm) instruction version were chosen.

This meant much of the TableGen machinery present for the assembly
matching was effectively being bypassed and worked around.

This patch makes the CC and CCOut operands optional which allows
the ASM matcher operate as it was designed and means we can avoid
doing some of the hacks done previously. This also adds the option
for the target to allow the prioritizing the smaller instruction
encodings as is required for Aarch32.

AlfieRichardsArm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -4052,6 +4073,66 @@ static MCRegister MatchRegisterName(StringRef Name);
 /// }
 
+static bool isDataTypeToken(StringRef Tok) {
+  return Tok == ".8" || Tok == ".16" || Tok == ".32" || Tok == ".64" ||
```
Done

AlfieRichardsArm commented:
Also, for quite a nice example of improved behavior from this change, see `llvm/test/MC/ARM/tMOVSr.s`.

In this case, the old behavior was that the de-facto wide/narrow decision meant this was parsed as a wide encoding which was then converted to narrow by a hacky bit of code that this test checks for. 

This test is changed by this patch to check that the narrow encoding is used because it was selected by the TableGen matcher and the hack wasn't used (I intend to go through these hacks and remove all those that are now unused after this patch is accepted).

s-barannikov commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -608,12 +627,18 @@ struct MatchableInfo {
+    // we must choose the smallest matching instruction.
+    if (Target.getPreferSmallerInstructions() && ResInstSize != RHS.ResInstSize)
+      return ResInstSize < RHS.ResInstSize;
+
```
I think this check should be tie-breaking rather than defining, and so it should be moved below, after we've compared the number of operands and their classes.
Also, does this check supersede the AVX/AVX512 resolution below?

AlfieRichardsArm replied:
It is my understanding this needs to be here for ARM. The architecture reference manual states: 
```
If neither .W nor .N is specified, the assembler can select either a 16-bit or 32-bit encoding. If both encoding
lengths are available, it must select a 16-bit encoding.
```
There are cases (such as thumb 1 instructions with a CCOut operand whereas thumb 2 doesnt) where the smaller instruction has more operands, so choosing that first would make this non-compliant. 

Either way this patch has been moved to https://github.com/llvm/llvm-project/pull/83587

s-barannikov commented on llvm/lib/Target/ARM/ARM.td:
```diff
@@ -1747,4 +1747,5 @@ def ARM : Target {
   let AssemblyParsers = [ARMAsmParser];
   let AssemblyParserVariants = [ARMAsmParserVariant];
   let AllowRegisterRenaming = 1;
+  let PreferSmallerInstructions= true;
```
Should it be a member of AsmParser instead?

AlfieRichardsArm replied:
Ah yeah that would make sense, I am happy to move it.

AlfieRichardsArm replied:
Will continue this in https://github.com/llvm/llvm-project/pull/83587

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -4053,6 +4073,63 @@ static MCRegister MatchRegisterName(StringRef Name);
 
 /// }
 
+static const DenseSet<StringRef> DataTypes{
```
This should be moved inside isDataTypeToken to avoid start-up initialization cost.

AlfieRichardsArm replied:
Done thank you

AlfieRichardsArm committed:
Make VPTPred optional also

s-barannikov commented on llvm/lib/Target/ARM/ARM.td:
```diff
@@ -1732,6 +1732,7 @@ def ARMAsmWriter : AsmWriter {
 
 def ARMAsmParser : AsmParser {
   bit ReportMultipleNearMisses = 1;
+  let PreferSmallerInstructions= true;
```
```suggestion
  let PreferSmallerInstructions = true;
```

AlfieRichardsArm replied:
Done

s-barannikov commented on llvm/lib/Target/ARM/ARMInstrFormats.td:
```diff
@@ -155,7 +155,12 @@ def iflags_op : Operand<i32> {
-def CondCodeOperand : AsmOperandClass { let Name = "CondCode"; }
+def CondCodeOperand : AsmOperandClass {
+  let Name = "CondCode";
+  let PredicateMethod = "isCondCode";
```
Looks redundant (the name of the method should be correctly inferred).

AlfieRichardsArm replied:
Done

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -5972,17 +6154,30 @@ ParseStatus ARMAsmParser::parseFPImm(OperandVector &Operands) {
     return ParseStatus::NoMatch;
 
-  Parser.Lex(); // Eat '#' or '$'.
+  if (Parser.getTok().getKind() == llvm::AsmToken::Hash ||
```
This is already checked above (6143-6144), isn't it?

AlfieRichardsArm replied:
Ahh yes, thank you

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -7050,29 +7098,24 @@ bool ARMAsmParser::ParseInstruction(ParseInstructionInfo &Info, StringRef Name,
 
   // Add the carry setting operand, if necessary.
-  if (CanAcceptCarrySet) {
+  if (CanAcceptCarrySet && CarrySetting != 0) {
```
```suggestion
  if (CanAcceptCarrySet && CarrySetting) {
```

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -7118,6 +7161,11 @@ bool ARMAsmParser::ParseInstruction(ParseInstructionInfo &Info, StringRef Name,
     }
   }
 
+  // This marks the end of the LHS Mnemonic operators
```
Statements should end with a period. Please make sure other comments are fixed as well.

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -12811,11 +12930,24 @@ unsigned ARMAsmParser::validateTargetOperandClass(MCParsedAsmOperand &AsmOp,
+      auto DPair = getDRegFromQReg(Op.getReg());
+      DPair = MRI->getMatchingSuperReg(
+          DPair, ARM::dsub_0, &ARMMCRegisterClasses[ARM::DPairRegClassID]);
+      Op.setVecListDPair(DPair);
```
This method shouldn't mutate the operand.
You may want to modify the `ARMOperand::isVecListDPair` instead to accept single register and `ARMOPerand::addXXXOperand` to render the register as a list.

AlfieRichardsArm replied:
Ah this makes sense, I didn't know where to do it. I followed the pattern of llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:1305 which does mutation at this point.
The functions needed for this conversion comes from the `MCRegisterInfo` object `MRI` which is a member of the `ARMAsmParser` class. This isn't available for at `ARMOperand::addXXXOperand` so this is quite challenging. I would like to avoid finding a way of plumbing this through if possible (I guess `ARMOperand` would need a reference to its parser which feels a bit gross) do you think it is wrong if this stays here or is it a matter of preference?

s-barannikov replied:
> The functions needed for this conversion comes from the MCRegisterInfo object MRI which is a member of the ARMAsmParser class. This isn't available for at ARMOperand::addXXXOperand so this is quite challenging. 

This is a known issue, and you correctly guessed one way to overcome it. It would be better if addXXXOperand methods were members of AsmParser instance, where everything that is needed is available, but they are not, and it is difficult to change now.

> do you think it is wrong if this stays here or is it a matter of preference?

I guess it can stay here, but "mutate it at your own risc". Once the operand is mutated, it can no longer be matched as non-mutated.

AlfieRichardsArm replied:
Ah the mutation not being reversible for future matching is a good point, I hadn't thought about that. I think in this case its okay as the circumstances where this conversion would happen are very specific but I will add a comment explaining at the very least.

AlfieRichardsArm committed:
Some smaller changes addressing comments. Additionally updated some definitions to
fix MCA tests.

AlfieRichardsArm commented:
Tests passing now (other than unrelated flir test failure on Windows)

s-barannikov commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -10984,6 +11087,12 @@ unsigned ARMAsmParser::checkTargetMatchPredicate(MCInst &Inst) {
+    // The second source operand must be the same register as the destination
+    // operand.
+    if (Inst.getOperand(0).getReg() != Inst.getOperand(3).getReg())
+      return Match_RequiresDestinationRegisterMatchASourceRegister;
```
This should be diagnosed by built-in tied operand checking. Does it not do that for some reason?

AlfieRichardsArm replied:
There's weirdness in this case. I can't quite remember why. I think because of `cvtThumbMultiply` being the custom match class that tries to swap `Rn` and `Rm` if needed to get them to match the normal method for this check doesn't work? 

Without this the matching of `muls r1, r2, r3` hits the error at llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:8036 when validating the narrow version so doesn't attempt to match the wide version.

AlfieRichardsArm replied:
Hmm yeah I can't see where in `build/lib/Target/ARM/ARMGenAsmMatcher.inc` such operand restrictions would be checked. Is this supported for instruction selection at all do you know?

s-barannikov replied:
It appears the code that checks the constraints is not emitted when ReportMultipleNearMisses is true:
```C++
  if (!ReportMultipleNearMisses)
    emitAsmTiedOperandConstraints(Target, Info, OS, HasOptionalOperands);
```
There is a comment elsewhere:
```
      // FIXME: The tied operands checking is not yet integrated with the
      // framework for reporting multiple near misses.
```
In any case, it seems that if an instruction has a custom converter, the built-in tied operands checking is not performed. 

PS
There is `Match_InvalidTiedOperand` that would be used if tied operand checking was emitted.

AlfieRichardsArm replied:
Ah great, thank you. That's a bit of a pain but for now I will add a `FIXME` explaining. I'll switch to using `Match_InvalidTiedOperand` also.

AlfieRichardsArm replied:
Done

s-barannikov commented on llvm/lib/Target/ARM/ARMInstrThumb.td:
```diff
@@ -1768,7 +1768,7 @@ let isBranch = 1, isTerminator = 1, isBarrier = 1, isIndirectBranch = 1 in {
 
 // In Thumb1, "nop" is encoded as a "mov r8, r8". Technically, the bf00
 // encoding is available on ARMv6K, but we don't differentiate that finely.
```
Should the comment be updated?

AlfieRichardsArm replied:
No, I don't think so. The `mov r8, r8` is still used for Thumb1. I added other code to make sure its not used in cases with the bf00 encoding available.

AlfieRichardsArm replied:
We also still dont differentiate for ARMv6k

AlfieRichardsArm committed:
Remove unecessary and outdated `tMUL` disgnosis.

AlfieRichardsArm committed:
Minor comment fixups.

AlfieRichardsArm commented:
Apologies for all the pushes updates today, the tests are all passing on my machine but its getting issues on the CI runner.

AlfieRichardsArm commented:
Looks like the test failures were coming from ambiguous matchables being sorted differently. I have made a new patch that disambiguates these cases here: https://github.com/llvm/llvm-project/pull/84483

s-barannikov commented:
> Looks like the test failures were coming from ambiguous matchables being sorted differently.

Why were they sorted differently?

AlfieRichardsArm commented:
Ah, looking at file that caused the failure (llvm/test/MC/ARM/negative-immediates-fail.s) we only previously checked for 1 of the errors. So if they were in different orders it wouldn't have caused a failure. So this reordering was probably happening previously but wasn't tested for.

AlfieRichardsArm commented:
Hmmm, the same problem remains irritatingly. There are cases of matchables with the same operands, features, but _different_ result instructions, which cause different diagnostics....... 
So I'm adding a sort for that also. To aforementioned commit.

AlfieRichardsArm committed:
Fix tests

AlfieRichardsArm committed:
Fix newly added VLLDM code

AlfieRichardsArm commented:
@s-barannikov @kosarev With the latest fix to https://github.com/llvm/llvm-project/pull/83587 this should now pass and be ready to merge

AlfieRichardsArm commented:
@statham-arm if you have time could you give it a brief look over. Its changed quite a bit since you last saw it.

s-barannikov commented on llvm/utils/TableGen/AsmMatcherEmitter.cpp:
```diff
@@ -3253,7 +3276,7 @@ void AsmMatcherEmitter::run(raw_ostream &OS) {
         const MatchableInfo &B = **J;
 
-        if (A.couldMatchAmbiguouslyWith(B)) {
+        if (A.couldMatchAmbiguouslyWith(B, Target)) {
```
```suggestion
        bool PreferSmallerInstructions =
            AsmParser->getValueAsBit("PreferSmallerInstructions");
        if (A.couldMatchAmbiguouslyWith(B, PreferSmallerInstructions)) {
```
Same for shouldBeMatchedBefore above.

AlfieRichardsArm replied:
Ah yes, I agree this is a good idea. I will address in https://github.com/llvm/llvm-project/pull/83587.

s-barannikov reviewed (approved):
Generic changes look good to me. I'm not qulified to review ARM-specific changes.

statham-arm commented on llvm/include/llvm/Target/Target.td:
```diff
@@ -1564,6 +1562,10 @@ class AsmParser {
   bit CallCustomParserForAllOperands = false;
+
+  // PreferSmallerInstructions - Should the assembly matcher prefer the smaller
+  // instructions.
```
I would add to this comment

(a) that this flag is useful for Arm (so that people who only know other architectures have an idea why it needs to be here in the first place)

(b) what the preference order is if this flag is false. (Surely the only reason this flag exists, instead of just being always-true, is so that somebody can deliberately set it to false, in which case they have some reason to want a different preference order, and will need to know how to get it.)

AlfieRichardsArm replied:
Addressing in https://github.com/llvm/llvm-project/pull/83587

AlfieRichardsArm replied:
Yes I agree with @kosarev also. The ARM backend has a massive number of ambiguous matchables (some are exact duplicate aliases, some are meaningfully different) that we rely on the stable sort to not change the order of. 
I have a patch to remove a handful of completely redundant matchables I'll clean up and send upstream at some point. 

As discussed in https://github.com/llvm/llvm-project/pull/84483 I do think there is a case to order by predicates (similar to current sorting by register classes), however there isn't a good way to do this that I can see.

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -1347,6 +1368,14 @@ class ARMOperand : public MCParsedAsmOperand {
   }
+  bool isDReg() const {
+    return isReg() &&
+           ARMMCRegisterClasses[ARM::QPRRegClassID].contains(Reg.RegNum);
```
This looks identical to `isQReg` below – it still checks against `QPRRegClassID`. Shouldn't it have a different id?

AlfieRichardsArm replied:
Whoops, thank you! I am surprised this didn't catch any tests.

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -4054,6 +4099,65 @@ static MCRegister MatchRegisterName(StringRef Name);
+      ++MnemonicOpsEndInd;
+  }
+
+  // In some circumstances the code code moves to the right
```
I guess "code code" is a braino for "condition code"

AlfieRichardsArm replied:
I think it was bad muscle memory for `cond code` which I have typed quite a bit in the last few weeks :).
Thank you for catching.

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -4054,6 +4099,65 @@ static MCRegister MatchRegisterName(StringRef Name);
+      RHSCondCode = true;
+      MnemonicOpsEndInd++;
+    }
+    // Special case for it instructions which have a condition code on the RHS
```
Copy-paste error? This comment is identical to the one five lines up, but surely can't be an accurate description of _both_ clauses of this if.

AlfieRichardsArm replied:
While I have been known to create such `if` structures in the past, this is indeed not one. Thank you.

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -4054,6 +4099,65 @@ static MCRegister MatchRegisterName(StringRef Name);
+                 isDataTypeToken(Op.getToken()))) {
+      // In the mnemonic operators the cond code must always precede the data
+      // type. So we can now safely assume any subsequent cond code is on the
+      // RHS. As is the cdase for VCMP and VPT.
```
(typo, while I'm here anyway: cdase → case)

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -6886,19 +6904,17 @@ void ARMAsmParser::fixupGNULDRDAlias(StringRef Mnemonic,
-  size_t NumPredOps = isPredicable ? 1 : 0;
 
-  if (Operands.size() <= 3 + NumPredOps)
+  if (Operands.size() < 2 + MnemonicOpsEndInd)
```
Should this be a `<=` instead of `<`? Further down, line 6952, you're dereferencing `Operands[MnemonicOpsEndInd + 2]`, which would be one beyond the end if `Operands.size()` was _exactly_ the value you're comparing it with here.

(The previous code is `<=`, which was what caught my eye)

AlfieRichardsArm replied:
Ahh yup, thank you

statham-arm commented on llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:
```diff
@@ -7398,71 +7424,83 @@ static bool instIsBreakpoint(const MCInst &Inst) {
                                        const OperandVector &Operands,
-                                       unsigned ListNo, bool IsARPop) {
-  const ARMOperand &Op = static_cast<const ARMOperand &>(*Operands[ListNo]);
-  bool HasWritebackToken = Op.isToken() && Op.getToken() == "!";
```
I didn't understand why the new version of this function doesn't need to take account of whether there's a writeback token. In this version, we're checking for one, and using it to adjust the index of the token on which we report an error. The replacement code doens't check.

Line 7994 seems to be checking for a writeback token to set `RegStart`, but it's then not passing the result to this function. So how is the operand index not going to be off by one depending on writeback? Are there tests somewhere demonstrating that this still works?

AlfieRichardsArm replied:
The only purpose of checking for the writeback token was to position position the error at the register list location. I made a function (`getRegListInd`) that instead finds the register list operand instead as I thought it more understandable. This then doesn't need to check the writeback as it finds the index a different way.

AlfieRichardsArm replied:
I will use the same function for line 7994 to be consistent.

statham-arm replied:
Ah, yes, now I see the call that's looking up the index, thanks.

statham-arm reviewed (commented):
This looks generally good to me, but I spotted a few nits :-)

AlfieRichardsArm committed:
Address comments

statham-arm reviewed (approved):
Thanks for the fixes. LGTM from my side now.

ostannard reviewed (approved):
This LGTM too.

mstorsjo commented:
This change broke building FFmpeg for arm targets. Assembling https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/arm/vp8_armv6.S#L73-L75 produces the following error:
```
src/libavcodec/arm/vp8_armv6.S:74:27: error: asr operand expected.
        pkhtbne r11, r11, r11, asr #16
                          ^
src/libavcodec/arm/vp8_armv6.S:75:9: error: incorrect condition in IT block; got 'al', but expected 'ne'
        ldr r8, [r0, #16] @ code_word
        ^
```

It can be reproduced with the following standalone test snippet:
```assembly
        .syntax unified
        @.thumb

        it ne
        pkhtbne r11, r11, r11, asr #16
        nop
```
Built like this:
```console
$ clang -target armv7-linux-gnueabihf -c repro.s
repro.s:5:27: error: asr operand expected.
        pkhtbne r11, r11, r11, asr #16
                          ^
repro.s:6:9: error: incorrect condition in IT block; got 'al', but expected 'ne'
        nop
        ^
```

Please consider reverting this if you can't come up with a fix during the day. (Or I can push a revert myself.)

Also, as a side note, from the merged git commit, I see that you have your email address hidden on github. Apparently the bot didn't make a comment about that for some reason. So, from https://github.com/llvm/llvm-project/blob/main/.github/workflows/email-check.yaml, please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account, and see [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

AlfieRichardsArm commented:
@mstorsjo I have a fix I will add you to shortly. Thank you for finding this. 
Do you have the compile command for the whole file? I would like to check that the file now compiles.

mstorsjo commented:
> @mstorsjo I have a fix I will add you to shortly. Thank you for finding this. Do you have the compile command for the whole file? I would like to check that the file now compiles.

I tested, the whole file, and the rest of the project, builds fine now - thanks! (It usually requires you to run configure for the project, to generate a suitable `config.h` for the target, so providing a single command for the full file isn't easy.)

dyung commented:
Hi @AlfieRichardsArm we are seeing a case with inline asm that no longer compiles after your change. Consider the following code:
```c
void a() {
  __asm("strexd r0, r2, [r3]");
  __asm("ldrexd r0, [r2]");
}
```
Prior to your change, this compiled successfully:
```
$ ~/src/upstream/3e6db602918435b6a5ac476f63f8b259e7e73af4-linux/bin/clang --target=armv7a-unknown-eabi -c min2.c
```
But after your change, the compiler is now emitting errors for the same code:
```
$ ~/src/upstream/295cdd5c3dbd14406bf9cce01e3dfd787fb1ddda-linux/bin/clang --target=armv7a-unknown-eabi -c min2.c
min2.c:2:9: error: invalid operand for instruction
    2 |   __asm("strexd r0, r2, [r3]");
      |         ^
<inline asm>:1:13: note: instantiated into assembly here
    1 |         strexd r0, r2, [r3]
      |                    ^
min2.c:3:9: error: invalid operand for instruction
    3 |   __asm("ldrexd r0, [r2]");
      |         ^
<inline asm>:1:9: note: instantiated into assembly here
    1 |         ldrexd r0, [r2]
      |                ^
2 errors generated.
```
Is this expected?

AlfieRichardsArm commented:
Hi @dyung,
This is likely a regression that there weren't tests for. I will look into this urgently tomorrow.
Alfie

AlfieRichardsArm commented:
Hi @dyung,
Looking at the architecture reference manual (https://developer.arm.com/documentation/ddi0597/2023-09/?lang=en) I dont recognise these encodings.

Perhaps we previously parsed these but shouldn't have? (For me it causes a crash on old builds?).

In this case they would both need an extra operands. eg. 
```
        strexd  r0, r2, r3, [r3]                @ encoding: [0x90,0x0f,0xa3,0xe1]
        ldrexd  r0, r1, [r2]                    @ encoding: [0x9f,0x0f,0xb2,0xe1]
```
(I'm guessing at the missing registers)

Is the source for the project which these came from public?

dyung commented:
> Hi @dyung, Looking at the architecture reference manual (https://developer.arm.com/documentation/ddi0597/2023-09/?lang=en) I dont recognise these encodings.
> 
> Perhaps we previously parsed these but shouldn't have? (For me it causes a crash on old builds?).
> 
> In this case they would both need an extra operands. eg.
> 
> ```
>         strexd  r0, r2, r3, [r3]                @ encoding: [0x90,0x0f,0xa3,0xe1]
>         ldrexd  r0, r1, [r2]                    @ encoding: [0x9f,0x0f,0xb2,0xe1]
> ```
> 
> (I'm guessing at the missing registers)
> 
> Is the source for the project which these came from public?

The code is unfortunately not from a public codebase.

I am told that it is a short form of essentially the longer form you posted. If you try earlier versions of clang and compile directly to an object file and disassemble it, I believe it generated the assembly you mentioned. (I also noticed that earlier versions of clang hit an assertion failure when generating the assembly)

AlfieRichardsArm commented:
Ah okay, very strange. I cant find any references to this in specifications or docs. Do you have any documentation for it? 

I would be happy to add some aliases for this alternate form if it's correct behavior, but I can't seem to find documentation for it or references to it before this code.

smithp35 commented:
It does look like GCC will accept these alternative forms. I expect that this will have been an unofficial/undocumented bit of GCC behaviour that clang has copied.

AlfieRichardsArm commented:
@smithp35 Ah okay, in that case shall I re-add this behavior and add some tests? 
I dislike the way `strexd` and `ldrexd` is handled so could rewrite this to handle this case and have better diagnostics.

smithp35 commented:
I think adding that behaviour back makes sense as existing implementations make a defacto standard. Ideally we can trace this back and when we add a test we can say it is for compatibility with GCC.

smithp35 commented:
Yes, all I've done is just plug the example on GCC. If we're lucky there will be a comment in the source code or commit message for GCC, or clang saying that it has copied GCC.

There will be a limit to what we can do, sometimes there won't be anything written down that can be easily found.

woruyu reviewed (commented):
Ask a question!
